### ISSUE TITLE
Rats [4.0.0] Can't rejoin a world

### ISSUE THREAD

**Issue by PokeMaserJ** [Alex-the-666/Rats#183]
So with upgrading forge to 28.21 I was able to create worlds with Rats, but if I left and tried to rejoin the world, I'd be stuck at preparing spawn. I can post the logs if needed.

**Comment by UnseenAcademical** [comment]
Most recent update didn't fix the problem, still freezing at 100% upon trying to rejoin a single player world in 14.4

**Comment by Peach774** [comment]
Also having this issue - it doesn't even error - it either sits at 100% or it sends me back to the main menu.

**Comment by drewbit** [comment]
Having the issue as well on the latest version of everything. (Rats 4.0.1 and Forge 28.2.3)

**Comment by Lhruggles** [comment]
I Also have this. I have tried everything and it still won't work. Apparently only some people have the issue though. Very strange.

**Comment by JakeMoe** [comment]
I had posted this about two hours ago on the mod's comments, but figured it probably belongs here.

I had the same issue with a small modpack I made, running in singleplayer (on Linux with minecraft-launcher); it'd get to 100% and just hang.  I thought it might be something to do with running the whole thing on my computer, so I made a 1.14.4 server modpack and started it up on the system I usually run my own modpack on.  While I can create the world, join it, and start to play, after I've disconnected, I can't get back; I get ServerHangWatchdog detecting a single tick took 60.00 seconds, should max be 0.05, and it shuts down the server.

In the crashlogs, they have in the first section:

```
java.lang.Error: ServerHangWatchdog detected that a single server tick took 60.00 seconds (should be max 0.05)
        at sun.misc.Unsafe.park(Native Method) ~[?:1.8.0_201] {}
        at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215) ~[?:1.8.0_201] {}
        at net.minecraft.util.concurrent.ThreadTaskExecutor.func_223705_bi(SourceFile:139) ~[?:?] {re:classloading,pl:accesstransformer:B}
        at net.minecraft.util.concurrent.ThreadTaskExecutor.func_213161_c(SourceFile:129) ~[?:?] {re:classloading,pl:accesstransformer:B}
        at net.minecraft.world.server.ServerChunkProvider.func_212849_a_(SourceFile:135) ~[?:?] {re:classloading,pl:accesstransformer:B}
        at net.minecraft.world.chunk.AbstractChunkProvider.func_217205_a(SourceFile:16) ~[?:?] {re:classloading}
        at net.minecraft.world.World.getBiomeBody(World.java:111) ~[?:?] {re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
        at net.minecraftforge.common.extensions.IForgeDimension.getBiome(IForgeDimension.java:164) ~[?:?] {re:classloading,pl:runtimedistcleaner:A}
        at net.minecraft.world.World.func_180494_b(World.java:106) ~[?:?] {re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
        at com.github.alexthe666.rats.server.events.CommonEvents.onEntityJoinWorld(CommonEvents.java:149) ~[?:4.0.1] {re:classloading}
        at net.minecraftforge.eventbus.ASMEventHandler_184_CommonEvents_onEntityJoinWorld_EntityJoinWorldEvent.invoke(.dynamic) ~[?:?] {}
        at net.minecraftforge.eventbus.ASMEventHandler.invoke(ASMEventHandler.java:80) ~[eventbus-1.0.0-service.jar:?] {}
        at net.minecraftforge.eventbus.EventBus.post(EventBus.java:258) ~[eventbus-1.0.0-service.jar:?] {}
        at net.minecraft.world.server.ServerWorld.func_217440_f(ServerWorld.java:904) ~[?:?] {re:classloading,xf:fml:quark:change-sleeping-player-count,xf:fml:quark:add-rave-hook,pl:runtimedistcleaner:A}
        at net.minecraft.world.server.ChunkManager.lambda$null$24(ChunkManager.java:577) ~[?:?] {re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
        at net.minecraft.world.server.ChunkManager$$Lambda$7787/261606681.apply(Unknown Source) ~[?:?] {}
        at com.mojang.datafixers.util.Either.lambda$mapLeft$0(Either.java:162) ~[?:?] {re:classloading}
        at com.mojang.datafixers.util.Either$$Lambda$7788/1212879725.apply(Unknown Source) ~[?:?] {}
        at com.mojang.datafixers.util.Either$Left.map(Either.java:38) ~[?:?] {re:classloading}
        at com.mojang.datafixers.util.Either.mapLeft(Either.java:162) ~[?:?] {re:classloading}
        at net.minecraft.world.server.ChunkManager.lambda$func_219200_b$25(ChunkManager.java:554) ~[?:?] {re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
        at net.minecraft.world.server.ChunkManager$$Lambda$7783/758994114.apply(Unknown Source) ~[?:?] {}
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602) ~[?:1.8.0_201] {}
        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577) ~[?:1.8.0_201] {}
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442) ~[?:1.8.0_201] {}
        at net.minecraft.world.chunk.ChunkTaskPriorityQueueSorter.func_219083_b(SourceFile:58) ~[?:?] {re:classloading}
        at net.minecraft.world.chunk.ChunkTaskPriorityQueueSorter$$Lambda$7736/1201306065.run(Unknown Source) ~[?:?] {}
        at net.minecraft.util.concurrent.ThreadTaskExecutor.func_213166_h(SourceFile:144) ~[?:?] {re:classloading,pl:accesstransformer:B}
        at net.minecraft.util.concurrent.ThreadTaskExecutor.func_213168_p(SourceFile:118) ~[?:?] {re:classloading,pl:accesstransformer:B}
        at net.minecraft.world.server.ServerChunkProvider$ChunkExecutor.func_213168_p(SourceFile:550) ~[?:?] {re:classloading}
        at net.minecraft.util.concurrent.ThreadTaskExecutor.func_213161_c(SourceFile:127) ~[?:?] {re:classloading,pl:accesstransformer:B}
        at net.minecraft.world.server.ServerChunkProvider.func_212849_a_(SourceFile:135) ~[?:?] {re:classloading,pl:accesstransformer:B}
        at net.minecraft.world.World.func_217353_a(World.java:164) ~[?:?] {re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
        at net.minecraft.world.IWorldReader.func_217348_a(IWorldReader.java:94) ~[?:?] {re:classloading}
        at net.minecraft.world.World.func_212866_a_(World.java:160) ~[?:?] {re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
        at net.minecraft.entity.Entity.func_70080_a(Entity.java:1149) ~[?:?] {re:classloading,pl:accesstransformer:B,xf:fml:quark:generic-update-hook,xf:fml:quark:override-collision,pl:runtimedistcleaner:A}
        at net.minecraft.network.play.ServerPlayNetHandler.func_175089_a(ServerPlayNetHandler.java:791) ~[?:?] {re:classloading,pl:accesstransformer:B}
        at net.minecraft.network.play.ServerPlayNetHandler.func_147364_a(ServerPlayNetHandler.java:776) ~[?:?] {re:classloading,pl:accesstransformer:B}
        at net.minecraft.server.management.PlayerList.func_72355_a(PlayerList.java:155) ~[?:?] {re:classloading,pl:runtimedistcleaner:A}
        at shadows.fastbench.net.HijackedDedicatedPlayerList.func_72355_a(HijackedDedicatedPlayerList.java:18) ~[?:2.1.4] {re:classloading}
        at net.minecraft.network.login.ServerLoginNetHandler.func_147326_c(ServerLoginNetHandler.java:116) ~[?:?] {re:classloading}
        at net.minecraft.network.login.ServerLoginNetHandler.func_73660_a(ServerLoginNetHandler.java:63) ~[?:?] {re:classloading}
        at net.minecraft.network.NetworkManager.func_74428_b(NetworkManager.java:224) ~[?:?] {re:classloading,pl:runtimedistcleaner:A}
        at net.minecraft.network.NetworkSystem.func_151269_c(NetworkSystem.java:135) ~[?:?] {re:classloading,pl:runtimedistcleaner:A}
        at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:846) ~[?:?] {re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
        at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:324) ~[?:?] {re:classloading,pl:accesstransformer:B}
        at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:764) ~[?:?] {re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
        at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:622) ~[?:?] {re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_201] {}
```

The method onEntityJoinWorld seemed to match what I was seeing; a hang while trying to join the world.  So I renamed the Rats JAR to rats.blah.jar.temp on both client and server, restarted them, and I can join fine now.  I just don't have any rats.  :-(

The line mentioned above seems to be pointing to an if clause, checking if the entity joining is a skeleton or zombie, and if it's in a jungle, and if so, use RNG to maybe spawn an archeologist's hat on it.  For testing, I commented out that whole if statement and recompiled, and I'm now able to rejoin both single-player worlds and multiplayer servers.  No idea why that check would cause the thread to hang, though.

Full server crash log is attached, FWIW

[crash-2020-03-07_16.13.56-server.txt](https://github.com/Alex-the-666/Rats/files/4301616/crash-2020-03-07_16.13.56-server.txt)

**Comment by Lhruggles** [comment]
wait so HOW did you get it to work

**Comment by JakeMoe** [comment]
TL:DR - it doesn't "work", I'm providing a bunch of info I found in the hope it'll help the mod author fix it properly. I know a bit of Java, and have done some Spigot and Sponge plugins in the past, but haven't touched Forge plugins.

I haven't found a "fix" that makes it "work"; I found that if I modified the source code in a small way, recompiled it, and added that mod file to my loaded mods, I can log in.  However, I've basically bypassed a small part of the code to see if it would make the login / join issue go away, so it really isn't 100% working properly.

I did some more tests on it yesterday, and found an alternate way of checking biomes, from LexManos: https://www.minecraftforge.net/forum/topic/56031-biomedictionary-causing-client-chunk-ticking/

So I tried this:

`if (event.getEntity() != null && (event.getEntity() instanceof AbstractSkeletonEntity || event.getEntity() instanceof ZombieEntity) && event.getWorld().getBiome(event.getEntity().getPosition()) == Biomes.JUNGLE) {`

but that had the same issue.  Changing it to remove the biome check altogether seems to work:

`if (event.getEntity() != null && (event.getEntity() instanceof AbstractSkeletonEntity || event.getEntity() instanceof ZombieEntity)) {`

HTH

**Comment by Lhruggles** [comment]
ok thanks

**Comment by Lhruggles** [comment]
I sure do hope Alex fixes it

**Comment by Alex-the-666** [comment]
should be fixed in 4.0.2, please go download it
